<a href="https://colab.research.google.com/github/ericodle/JP_BERT/blob/main/BERT_adverb_setup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

BERT for Japanese Adverb Predicgtion - Setup.
This is a general BERT setup for the Japanese langauge translated and adapted from the work of Yutaro Ogawa (https://github.com/YutaroOgawa/BERT_Japanese_Google_Colaboratory). 
The intent of this setup is to familiarize readers with the basics of our Japanese adverb prediction project. 

Install MeCab.

In [1]:
!apt install aptitude swig
!aptitude install mecab libmecab-dev mecab-ipadic-utf8 git make curl xz-utils file -y

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  aptitude-common libcgi-fast-perl libcgi-pm-perl libclass-accessor-perl
  libcwidget3v5 libencode-locale-perl libfcgi-perl libhtml-parser-perl
  libhtml-tagset-perl libhttp-date-perl libhttp-message-perl libio-html-perl
  libio-string-perl liblwp-mediatypes-perl libparse-debianchangelog-perl
  libsigc++-2.0-0v5 libsub-name-perl libtimedate-perl liburi-perl libxapian30
  swig3.0
Suggested packages:
  aptitude-doc-en | aptitude-doc apt-xapian-index debtags tasksel
  libcwidget-dev libdata-dump-perl libhtml-template-perl libxml-simple-perl
  libwww-perl xapian-tools swig-doc swig-examples swig3.0-examples swig3.0-doc
The following NEW packages will be installed:
  aptitude aptitude-common libcgi-fast-perl libcgi-pm-perl
  libclass-accessor-perl libcwidget3v5 libencode-locale-perl libfcgi-perl
  libhtml-parser-perl libhtml-tagset-perl libhttp

In [2]:
!pip install mecab-python3==0.996.6rc2

     |████████████████████████████████| 17.1 MB 21.8 MB/s 


Install the Neologd ipadic dictionary, as it contains newer words from the web.

In [3]:
!git clone --depth 1 https://github.com/neologd/mecab-ipadic-neologd.git
!echo yes | mecab-ipadic-neologd/bin/install-mecab-ipadic-neologd -n -a

Cloning into 'mecab-ipadic-neologd'...
remote: Enumerating objects: 75, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (74/74), done.
remote: Total 75 (delta 5), reused 54 (delta 0), pack-reused 0
Unpacking objects: 100% (75/75), done.
[install-mecab-ipadic-NEologd] : Start..
[install-mecab-ipadic-NEologd] : Check the existance of libraries
[install-mecab-ipadic-NEologd] :     find => ok
[install-mecab-ipadic-NEologd] :     sort => ok
[install-mecab-ipadic-NEologd] :     head => ok
[install-mecab-ipadic-NEologd] :     cut => ok
[install-mecab-ipadic-NEologd] :     egrep => ok
[install-mecab-ipadic-NEologd] :     mecab => ok
[install-mecab-ipadic-NEologd] :     mecab-config => ok
[install-mecab-ipadic-NEologd] :     make => ok
[install-mecab-ipadic-NEologd] :     curl => ok
[install-mecab-ipadic-NEologd] :     sed => ok
[install-mecab-ipadic-NEologd] :     cat => ok
[install-mecab-ipadic-NEologd] :     diff => ok
[install-mecab-ipadic-NEologd] :   

In [4]:
import subprocess

cmd='echo `mecab-config --dicdir`"/mecab-ipadic-neologd"'
path_neologd = (subprocess.Popen(cmd, stdout=subprocess.PIPE,
                           shell=True).communicate()[0]).decode('utf-8')

Check to make sure MeCab is working properly.

In [5]:
import MeCab

m=MeCab.Tagger("-Ochasen")

text = "私は機械学習が好きです。"

text_segmented = m.parse(text)
print(text_segmented)

私	ワタシ	私	名詞-代名詞-一般		
は	ハ	は	助詞-係助詞		
機械	キカイ	機械	名詞-一般		
学習	ガクシュウ	学習	名詞-サ変接続		
が	ガ	が	助詞-格助詞-一般		
好き	スキ	好き	名詞-形容動詞語幹		
です	デス	です	助動詞	特殊・デス	基本形
。	。	。	記号-句点		
EOS



In [6]:
m=MeCab.Tagger("-Owakati")
text_segmented = m.parse(text)
print(text_segmented)


私 は 機械 学習 が 好き です 。 



In [7]:
m=MeCab.Tagger("-Oyomi")
text_segmented = m.parse(text)
print(text_segmented)

ワタシハキカイガクシュウガスキデス。



Prepare a pre-trained Japanese BERT model

In [8]:


!pip install transformers==3.0.2



     |████████████████████████████████| 769 kB 26.6 MB/s 
     |████████████████████████████████| 3.0 MB 44.5 MB/s 
     |████████████████████████████████| 895 kB 47.3 MB/s 
     |████████████████████████████████| 1.2 MB 41.3 MB/s 


In [9]:
import torch
from transformers.modeling_bert import BertModel
from transformers.tokenization_bert_japanese import BertJapaneseTokenizer


Define the tokenizer. We will use the cl-Tohoku model and tokenizer.

In [10]:
tokenizer = BertJapaneseTokenizer.from_pretrained('cl-tohoku/bert-base-japanese-whole-word-masking')

Downloading:   0%|          | 0.00/258k [00:00<?, ?B/s]

Here you can see the details of our BERT model.

In [11]:
model = BertModel.from_pretrained('cl-tohoku/bert-base-japanese-whole-word-masking')
print(model)

Downloading:   0%|          | 0.00/479 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/445M [00:00<?, ?B/s]

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(32000, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0): BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          

Here you can see model details, such as the activation function of the hidden laters (gelu). Each vocabulary term is a 768-dimension vector. Up to 512 characters (subwords) can be handled at once.

In [14]:
from transformers import BertConfig

config_japanese = BertConfig.from_pretrained('cl-tohoku/bert-base-japanese-whole-word-masking')
print(config_japanese)

BertConfig {
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "tokenizer_class": "BertJapaneseTokenizer",
  "type_vocab_size": 2,
  "vocab_size": 32000
}



Here is a demonstration of how pytorch/BERT assign an id number to vocabulary so that math can be performed on words. Remember that each word is a 768-dimension vector, but BERT cannot handle so many numbers at one time. The character ”を”, for example, has an id of "811", which points to a list in the tokenizer dictionary that contains a 768-dimension vector representing ”を” in the Japanese language.

In [15]:
text1 = "会社をクビになった。"
text2 = "テレワークばかりでクビが痛い。"
text3 = "会社を解雇された。"

Note that the sentences start with a [CLS] token and end with a [SEP] token. This is the BERT format and is true for every input string.

In [16]:
input_ids1 = tokenizer.encode(text1, return_tensors='pt')  #pt stands for "pytorch"

print(tokenizer.convert_ids_to_tokens(input_ids1[0].tolist()))  # print the Japanese text
print(input_ids1)  # print the corresponding ids for the characters

['[CLS]', '会社', 'を', 'クビ', 'に', 'なっ', 'た', '。', '[SEP]']
tensor([[    2,   811,    11, 13700,     7,    58,    10,     8,     3]])


Repeat for text2

In [17]:
input_ids2 = tokenizer.encode(text2, return_tensors='pt')  

print(tokenizer.convert_ids_to_tokens(input_ids2[0].tolist()))  
print(input_ids2)  

['[CLS]', 'テレ', '##ワーク', 'ばかり', 'で', 'クビ', 'が', '痛', '##い', '。', '[SEP]']
tensor([[    2,  5521,  3118,  4027,    12, 13700,    14,  4897, 28457,     8,
             3]])


Repeat for text3

In [18]:
input_ids3 = tokenizer.encode(text3, return_tensors='pt') 

print(tokenizer.convert_ids_to_tokens(input_ids3[0].tolist()))  
print(input_ids3)

['[CLS]', '会社', 'を', '解雇', 'さ', 'れ', 'た', '。', '[SEP]']
tensor([[   2,  811,   11, 7279,   26,   20,   10,    8,    3]])


Next, input our text (converted to ids) into BERT. Let's try text1. As you can see, the result is a tensor (a matrix of matrices).

In [23]:
result1 = model(input_ids1)

print(result1)

(tensor([[[-0.0882,  0.1803, -0.0807,  ..., -0.4849,  0.1793, -0.0791],
         [ 0.5232,  0.2362,  0.1158,  ..., -0.3432, -0.0884, -0.6000],
         [-0.0379, -0.3802, -0.4018,  ..., -0.4543, -0.5445,  0.5767],
         ...,
         [ 0.0365, -0.0860,  0.1496,  ..., -0.4938,  0.3092, -0.0260],
         [-0.5299,  0.0150, -0.2162,  ..., -0.6310,  0.7568,  0.2828],
         [-0.7848, -0.0012, -0.4115,  ..., -0.6644,  0.8711,  0.1109]]],
       grad_fn=<NativeLayerNormBackward0>), tensor([[ 6.3560e-01,  2.2145e-01, -2.1823e-01, -1.8074e-01,  3.0972e-01,
          9.9889e-01,  7.9698e-03, -1.1052e-01, -6.8074e-02,  1.2303e-01,
         -9.6402e-01,  2.8340e-01,  6.6656e-02, -5.1850e-02, -1.0601e-01,
          2.5921e-01, -3.1930e-02,  1.0062e-01,  1.9708e-01,  2.2505e-01,
          3.1557e-01,  3.2451e-01, -3.5328e-01,  9.8977e-01,  9.5820e-01,
          2.8834e-01,  2.3717e-01, -1.6124e-01,  9.9954e-01,  1.3057e-01,
         -4.3924e-01,  4.8160e-03, -1.7704e-01,  1.7651e-01,  1.2831e

We can get more general information about our tensor result with with .shape() function. Also note that result1[0] represents the sequence output, and result1[1] represents the pooled output.

In [30]:
print(result1[0].shape)
print(result1[1].shape)


torch.Size([1, 9, 768])
torch.Size([1, 768])


Result1[0] contains 1 row, 9 columns... and 768 elements in a third dimension. "1" represents the number of batches, "9" represents the 9 ids that we fed into BERT from text1, and "768" represents the dimensions of each input id. 
Result1[1] contains 1 row and 768 colums (or rather, a list of 768 numbers). 

In [31]:
print(result1[0])

tensor([[[-0.0882,  0.1803, -0.0807,  ..., -0.4849,  0.1793, -0.0791],
         [ 0.5232,  0.2362,  0.1158,  ..., -0.3432, -0.0884, -0.6000],
         [-0.0379, -0.3802, -0.4018,  ..., -0.4543, -0.5445,  0.5767],
         ...,
         [ 0.0365, -0.0860,  0.1496,  ..., -0.4938,  0.3092, -0.0260],
         [-0.5299,  0.0150, -0.2162,  ..., -0.6310,  0.7568,  0.2828],
         [-0.7848, -0.0012, -0.4115,  ..., -0.6644,  0.8711,  0.1109]]],
       grad_fn=<NativeLayerNormBackward0>)


To make sure that our dimensions are correct, and to see the raw output from out BERT model, we can simply print the pooled output, which is result1[1]. Also note that the hyperbolic tangent function (tanh) is the activation function used in the final ouput layer.

In [32]:
print(result1[1])

tensor([[ 6.3560e-01,  2.2145e-01, -2.1823e-01, -1.8074e-01,  3.0972e-01,
          9.9889e-01,  7.9698e-03, -1.1052e-01, -6.8074e-02,  1.2303e-01,
         -9.6402e-01,  2.8340e-01,  6.6656e-02, -5.1850e-02, -1.0601e-01,
          2.5921e-01, -3.1930e-02,  1.0062e-01,  1.9708e-01,  2.2505e-01,
          3.1557e-01,  3.2451e-01, -3.5328e-01,  9.8977e-01,  9.5820e-01,
          2.8834e-01,  2.3717e-01, -1.6124e-01,  9.9954e-01,  1.3057e-01,
         -4.3924e-01,  4.8160e-03, -1.7704e-01,  1.7651e-01,  1.2831e-01,
          1.3220e-01, -4.9360e-02, -4.6955e-01, -7.8238e-02,  1.8543e-01,
         -1.1779e-01, -9.8242e-01,  6.2389e-01, -2.7589e-01, -6.9216e-01,
         -1.4458e-02,  1.9695e-01, -1.2863e-01, -1.8545e-01, -8.0568e-02,
         -3.0220e-01, -3.5999e-01,  6.0221e-01, -9.9994e-01, -3.2288e-01,
          1.6276e-01, -1.6985e-01, -1.0654e-01, -4.2879e-02,  9.6053e-01,
          8.6668e-02,  2.8381e-01,  2.5336e-01, -1.1414e-01,  4.9371e-02,
          2.8683e-01,  2.1428e-01,  2.

Now, let's feed text2 and text3 into BERT and get result2 and result3.

In [36]:
result2 = model(input_ids2)
result3 = model(input_ids3)

Each word in BERT is represented by a 768-dimension vector. Before we fed the text into BERT, those vectors were determined by the dictionary/tokenizer. After we fed the text through BERT, the 768 dimensions of each vector (word) was changed! 
Now, we can compare the outputs for each sentence and do some mathematical comparisons. 

Let's consider two words in our three sentences: "クビ" and "解雇". First, we will identify the target words in our three results.

In [39]:
word_vec1 = result1[0][0][3][:]  # ”クビ” in text1 (id #3)
word_vec2 = result2[0][0][5][:]  # ”クビ” in text2（id #5）
word_vec3 = result3[0][0][3][:]  # ”解雇” in text3（id #3）

We can confirm that the vectors for ”クビ” have changed based on the surrounding words in its sentences (in other words, the vectors have changed based on the BERT model and the input sentence context.)

In [45]:
print("Output 'クビ' from Text 1"+'\n')
print(word_vec1)
print("Output 'クビ' from Text 2"+'\n')
print(word_vec2)

Output 'クビ' from Text 1

tensor([ 1.2584e-01, -6.1721e-01, -6.1346e-01, -3.8816e-01,  1.9818e-01,
        -8.7479e-03,  2.6546e-01,  2.2833e-01,  1.6419e-01,  5.1053e-01,
         4.0090e-02,  7.3816e-01, -3.5953e-01, -2.8629e-02, -4.7689e-01,
        -4.7942e-02, -7.7567e-01,  3.9061e-01, -7.6744e-02, -1.8135e-01,
        -2.8891e-01, -3.1622e-01, -9.2472e-01, -3.5994e-02, -2.7289e-01,
        -2.7283e-01,  4.4309e-01,  4.3873e-01,  1.2904e-01, -1.2369e-01,
        -4.6541e-01,  6.5855e-01,  4.5617e-01,  4.3642e-01, -3.8470e-01,
        -6.7394e-02,  5.3484e-01, -3.8518e-01, -1.1368e-01, -8.5795e-01,
        -6.3291e-01,  5.7788e-01, -5.8622e-01, -3.9338e-02,  1.9570e-01,
        -8.6917e-02,  2.5688e-01, -1.4356e-01, -4.2090e-01, -2.8274e-01,
         3.9734e-02,  3.5462e-01, -2.5369e-02, -4.0318e-02,  3.7608e-02,
         3.6689e-01, -5.1964e-01,  2.1622e-01, -8.0713e-01,  1.3884e-01,
        -5.3604e-01, -3.9205e-01,  3.5426e-02,  4.8179e-01, -4.4411e-01,
        -5.7368e-02,  5.31

Now we want to answer the question: 

"Is the meaning of ”クビ” in Text 1 more similar to the meaning of ”クビ” in Text 2? Or, is the meaning of ”クビ” in Text 1 more similar to the meaning of ”解雇” in Text 3?"

We can answer this question many ways. One way is to use Cosine Similarity, which follows the formula: 

similarity=max(∥x1​∥2​⋅∥x2​∥2​,ϵ)x1​⋅x2​​

In [46]:
cos = torch.nn.CosineSimilarity(dim=0)
cos_sim_12 = cos(word_vec1, word_vec2)
cos_sim_13 = cos(word_vec1, word_vec3)

print(cos_sim_12)
print(cos_sim_13)

tensor(0.6647, grad_fn=<DivBackward0>)
tensor(0.7841, grad_fn=<DivBackward0>)


The cosine similarity between ”クビ” in text1 and text2 is 0.66, whereas the similarity between ”クビ” in text1 and ”解雇” in text3 is 0.78. 

Therefore, the meaning of ”クビ” in text1 and the meaning of ”解雇” in text3 is more similar than the meanings of ”クビ” in text1 and text2. 

This is correct, as the ”クビ” in text1 refers to getting fired from a job (”解雇”) not the anatomical "neck" meaning in text2.
